### PR TITLE
Improve caching of clients to reduce Azure Functions burps.

### PR DESCRIPTION
**Pull request by mitchdenny**
This PR adds a bunch of caching around the creation of Cosmos clients, and Azure DevOps clients used to make outbound HTTP requests. Current implementation was basically throwing away these objects on each call which means that we reach a threshold with Azure Functions fairly quickly where it wants to tear down the function instance to avoid reading max TCP connection limits.

This should help mitigate this. There is still an issue with cache stampeeding which I'm not happy about but before I try to come up with an elaborate solution for that with locking etc I want to make sure I am using caching correctly. But this change should fix these Azure Functions burps as they exist today.

Note ... I am deliberately not disposing of the VssConnection and various VssHttpClientBase objects when they are purged from the cache and instead relying on garbage collection because of the cache stampeed issue (I don't know for sure that there isn't an instance of the connection being relied on for a little longer than I expect).

Once the stampeed protection is better understood I'll also look at more explicitly managing the dependencies between cache entries such that if say the PAT changes that the entire chain of cache entries that rely on that are also torn down. But that is all in a future PR.

(this change is necessary to stop Pipeline Witness from potentially losing pipeline runs)

### CODE REVIEW

File: tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/Azure.Sdk.Tools.PipelineWitness.csproj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (JimSuplizio) [inline tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/Azure.Sdk.Tools.PipelineWitness.csproj:1]: nit: There's no actual change here

File: tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/RunProcessor.cs [hunk 2]
```diff
@@ -60,6 +60,41 @@ private Guid GetCollectionIdFromProject(TeamProject project)
...
         }
 
+        private async Task<VssConnection> GetVssConnectionAsync(string organization)
+        {
+            var vssConnectionCacheKey = $"{organization}_vssConnectionCacheKey";
+
+            var connection = await cache.GetOrCreateAsync(vssConnectionCacheKey, async (entry) =>
+            {
+                var azureDevOpsPersonalAccessToken = await GetAzureDevOpsPersonalAccessTokenAsync();
+                var credentials = new VssBasicCredential("nobody", azureDevOpsPersonalAccessToken);
```
Comment (JimSuplizio) [inline tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/RunProcessor.cs:70]: Should "nobody" be a constant somewhere?

Comment (mitchdenny) [reply tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/RunProcessor.cs:73]: Its never going to need to change, username is ignored by Azure DevOps for PAT-based auth.

File: tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/Azure.Sdk.Tools.PipelineWitness.csproj (hunk 1 above)
Comment (mitchdenny) [reply tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/Azure.Sdk.Tools.PipelineWitness.csproj]: Wonder if it is an encoding change.

**Pull request merged**